Ensure MEF initialization is completed for NugetPackage before used

erdembayar opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/10559

Regression? Last working version:

## Description

During VS experimental instance debugging it always throws exception due to MEF init is not completed yet. So we're fixing this with JTF aware Reentrant lock. It doesn't prematurely set `_initialized` to `true`, but wait until MEF init is complete.

Currently there's no any telemetry is emitted this situation. 
## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

erdembayar committed:
Check null reference before making call to SolutionManager.

nkolev92 commented on src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:
```diff
@@ -1160,14 +1160,18 @@ private void BeforeQueryStatusForAddPackageForSolutionDialog(object sender, Even
-                    !isConsoleBusy &&
-                    await SolutionManager.Value.DoesNuGetSupportsAnyProjectAsync();
+
+                if (SolutionManager != null)
```
Why is the solution manager null? 

If you look at the implementation of the method in question, it actually doesn't use the solution manager logic, it's really a method that could be converted to a static method if we really wanted to.

erdembayar replied:
> Why is the solution manager null?
> 

Problem very is similar to [PR#3889](https://github.com/NuGet/NuGet.Client/pull/3889) which I'm trying to fix timing bug, async init method `InitializeMEFAsync` still running but it already set[ `initialized = true` ](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L199), so `ShouldMEFBeInitialized` return false.

https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1149-L1152

Since MEF injection not complete yet for `IVsSolutionManager` type `SolutionManager` still null.
I only experience this during debugging. I can make same fix as [PR#3889](https://github.com/NuGet/NuGet.Client/pull/3889), but for now this check null thing is good enough so I want to keep it simple.

> If you look at the implementation of the method in question, it actually doesn't use the solution manager logic, it's really a method that could be converted to a static method if we really wanted to.

Are you referring below method? We can't convert to static method because [_asyncServiceProvider](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L383) need to be injected at [constructor](https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L161) first from another MEF injection. It gets ugly if we go that direction.

https://github.com/NuGet/NuGet.Client/blob/1729f1616e9b1812012bd4e0170ddffe1a2991ae/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs#L374-L383

nkolev92 replied:
There are ways to force MEF to satisfy the imports. 
We shouldn't be running into that race condition.

> Are you referring below method? We can't convert to static method because _asyncServiceProvider need to be injected at constructor first from another MEF injection. It gets ugly if we go that direction.


The NuGetPackage class itself is an async service provider.

erdembayar replied:
> at

Can you refer any examples how to force MEF to satisfy the imports? I prefer this one because it's minimal change and  doesn't change anything else.

nkolev92 replied:
The MEF initialization is actually supposed to happen: https://github.com/NuGet/NuGet.Client/blob/5ce9c44823a8930dd0bf57adbf5c32914d8e4ae3/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1149-L1152. 

> Since MEF injection not complete yet for IVsSolutionManager type SolutionManager still null.

I'm not confident your understand of the problem is correct. As such I don't think this is the best course of action at this point. 

Note that if SolutionManager is null, then the other problem you linked never comes into play since there's no object to be initialized.

erdembayar replied:
I wrapped MEF initializer with Reentrant lock and tested on my local it looks good, error disappears. After I remove it error reemerge.

donnie-msft reviewed (approved):
It looks like it opens once the SolutionManager is eventually initialized. I guess the command execution is queued up somewhere.

zivkan reviewed (approved):
1. This PR changes the shipping product, not just NuGet.Client repo development, so the PR title is misleading
2. While I agree with Nikolche that it would be nicer to have a more ideal solution, I think the PM UI button in the context menu being disabled for a longer time is an acceptable customer experience. it effectively becomes a performance problem, and if customers complain then we can treat it as a perf issue and work/schedule it appropriately.

zivkan commented:
> It looks like it opens once the SolutionManager is eventually initialized.

My guess was that the item in the context menu is disabled, and the customer has to "close" the context menu and keep re-opening it until it's finally not disabled.  It would be nice for Erick to explain the actual customer experience of this change.

donnie-msft commented:
> My guess was that the item in the context menu is disabled, and the customer has to "close" the context menu and keep re-opening it until it's finally not disabled.

If that's happening, I'd like to understand the impact as well @erdembayar.

My understanding of this bug is:
- The context menu is shown/enabled before initialization occurs. (PR doesn't change this).
- It's only observable when debugging (explains the issue/PR title). 
- Selecting `Manage NuGet Packages` in menu without a debugger attached, PMUI eventually opens (PR doesn't change this).

erdembayar commented:
> > My guess was that the item in the context menu is disabled, and the customer has to "close" the context menu and keep re-opening it until it's finally not disabled.
> 
> If that's happening, I'd like to understand the impact as well @erdembayar.
> 
> My understanding of this bug is:
> 
> * The context menu is shown/enabled before initialization occurs. (PR doesn't change this).
> * It's only observable when debugging (explains the issue/PR title).
> * Selecting `Manage NuGet Packages` in menu without a debugger attached, PMUI eventually opens (PR doesn't change this).

Yes, it doesn't change anything to end user. Only change noticeable is for people who is running experimental instance under debugger.

erdembayar commented:
> > It looks like it opens once the SolutionManager is eventually initialized.
> 
> My guess was that the item in the context menu is disabled, and the customer has to "close" the context menu and keep re-opening it until it's finally not disabled. It would be nice for Erick to explain the actual customer experience of this change.

Customers wouldn't experience any change. It only happens under debugger, it looks initialization code runs very slow under debugger which is causing this timing bug.

zivkan commented:
> Yes, it doesn't change anything to end user. Only change noticeable is for people who is running experimental instance under debugger.

The changes in the pull request don't make this obvious. How is it possible that this timing bug cannot possibly happen to a customer?

zivkan commented:
Alternatively, when this issue happens under debugging, does it generate any fault telemetry?  If so, have you checked fault telemetry to confirm that customers are never hitting it?

nkolev92 reviewed (changes requested):
I'm not sure we have full understanding of the root cause of the problem. 
Yes, it's null, but why is it null?

For all we know it could even be a bug outside of our code, but we should understand that first.

erdembayar committed:
Understand how often open PMUI exception due to SolutionManager not ready happen.

erdembayar commented:
> Alternatively, when this issue happens under debugging, does it generate any fault telemetry? If so, have you checked fault telemetry to confirm that customers are never hitting it?

Currently there is no telemetry to detect this one, I checked VS Telemetry monitor tool. Instead of trying to fix now I changed my code, now I added telemetry to detect it. Please check.

erdembayar commented:
> I'm not sure we have full understanding of the root cause of the problem.
> Yes, it's null, but why is it null?
> 
> For all we know it could even be a bug outside of our code, but we should understand that first.

Ok. Please check my other [comment](https://github.com/NuGet/NuGet.Client/pull/3891#issuecomment-777718575).

nkolev92 commented on src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:
```diff
@@ -1160,14 +1160,22 @@ private void BeforeQueryStatusForAddPackageForSolutionDialog(object sender, Even
+                        !isConsoleBusy &&
+                        await SolutionManager.Value.DoesNuGetSupportsAnyProjectAsync();
+                }
+                catch (Exception e)
```
The `BeforeQueryStatusForAddPackageForSolutionDialog` handler is something that's called from another component, do they have the telemetry?

erdembayar replied:
It's called from [line#315](https://github.com/NuGet/NuGet.Client/blob/5ce9c44823a8930dd0bf57adbf5c32914d8e4ae3/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L315) and that method `AddMenuCommandHandlersAsync` itself is called from line 159.

https://github.com/NuGet/NuGet.Client/blob/5ce9c44823a8930dd0bf57adbf5c32914d8e4ae3/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L151-L160

I don't see any specific telemetries for calling methods, I can see below events right before exception happen, not sure if it helps:

![image](https://user-images.githubusercontent.com/8766776/107693685-4d565e00-6c63-11eb-914b-3aa40f78464d.png)

Last one details:

![image](https://user-images.githubusercontent.com/8766776/107693098-86da9980-6c62-11eb-9f72-2e2dd54a6cda.png)

nkolev92 replied:
Thanks for looking at that. 

What are you trying to get through the telemetry? It has the same concern as the earlier implementation. We should try to fix it rather than workaround it.

erdembayar replied:
> Thanks for looking at that.
> 
> What are you trying to get through the telemetry? It has the same concern as the earlier implementation. We should try to fix it rather than workaround it.

I thought [we're interested](https://github.com/NuGet/NuGet.Client/pull/3891#issuecomment-776967793) in how it affect customer before fix it. 
Ok. In that case I'll rather fix it than telemetry and workaround.

nkolev92 replied:
I think @zivkan's point was that we should try to estimate the impact because you had claimed that it doesn't affect customers, so he was just asking whether we have more than just a guess there.

zivkan replied:
I think the issue looks like a real product bug that can affect customers, hence I disagree with claims such as "Customers wouldn't experience any change".

erdembayar committed:
Revert "Understand how often open PMUI exception due to SolutionManager not ready happen."

This reverts commit e48747c28315a41360a1e52325b99b8b6de21b96.

erdembayar committed:
Implement Reentrant lock.

erdembayar commented:
@nkolev92 @zivkan @donnie-msft 
Please review again. I used reentrant lock make sure MEF initialization is complete. With this implementation even I try to open PMUI right away null exception wouldn't happen. Once I remove it problem persist again.

nkolev92 commented on src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs:
```diff
@@ -196,53 +197,61 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
-            _initialized = true;
+            await _semaphore.ExecuteAsync(async () =>
+            {
+                if (_initialized)
```
Have you considered moving this check outside of the semaphore?

erdembayar replied:
@nkolev92 
I thought about it but ever time before `InitializeMEFAsync` called it also calls `ShouldMEFBeInitialized`

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L597-L600

Once MEF is completed `_initialized` will be checked in synchronously context below:

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1230-L1236

But above check is only matters during MEF init process. If anything is still waiting MEF to be completed then they can leave this context without re-starting init process.

nkolev92 reviewed (approved):
Nice job! Glad we got to the root cause of the problem.

Consider cleaning up the PR description to talk about the fix rather than the issue.
